### PR TITLE
contrib/utilities/update-copyright.sh: several improvements

### PR DESCRIPTION
**Pull request by tamiko**
Separated from #16686 to facilitate easier review.

 - make lookup of last modification year more robust
 - add a --pedantic mode that greps the first year a file was modified from git log
 - restructure processing into a separate bash function and add parallel processing structure
 - run script on all files of the repository

### CODE REVIEW

**Comment by tamiko** [comment]
@masterleinad I will run this script tomorrow on the header branch and verify that this indeed fixes all date ranges.

**Comment by tamiko** [comment]
@tjhei *ping*

**Review by tjhei — APPROVED** [review]
Looks good, but, wow, what a complex machinery you have built.

File: contrib/utilities/update-copyright.sh [hunk 1]
```diff
@@ -24,51 +25,162 @@ if test ! -d source -o ! -d include -o ! -d examples ; then
...
+    head -n 1 | \
+    perl -p -e 's/^(\d\d\d\d)-.*/\1/g;'`
+
+  #
+  # It should not happen, that the grep removes all 3 most recent commits
+  # simultaneously but if it does then run the git log command again with
+  # full history:
+  #
+
+  [ -z "$last_year" ] && last_year=`git log --date=short --format="format:%cd %s" ${file} | \
```
Comment (bangerth) [inline contrib/utilities/update-copyright.sh:88]: Out of curiosity, why didn't you just write this as `if test -z "$last_year" ; then ...`?

Comment (tamiko) [reply contrib/utilities/update-copyright.sh:88]: This is really purely a stylistic choice:
 - `test` and `[` are the same shell built-in. And `[ -z ... ]` looks better than `test -z ...`. I should have probably used the extended test pattern `[[ ... ]]` consistently...
 - I tend to use the `[[ ... ]] && VARIABLE="..."` a lot for simple conditional variable assignments. That's somewhat a bash coding style coming from Gentoo ebuilds. But yes, we could have used an `if` statement as well.

Comment (bangerth) [reply contrib/utilities/update-copyright.sh:88]: I see. It's perhaps a bit harder to read, but I'm ok with the stylistic choice.

**Pull request merged**